File: app/src/main/java/com/yuyin/demo/FilesManager.java
package com.yuyin.demo;

import android.content.Intent;
import android.os.Bundle;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import android.os.Environment;
import android.provider.DocumentsContract;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.yuyin.demo.databinding.FragmentFilesManagerBinding;

import java.io.File;
import java.io.FilenameFilter;
import java.io.IOException;
import java.lang.ref.PhantomReference;
import java.util.ArrayList;


public class FilesManager extends Fragment {

    private FragmentFilesManagerBinding binding;
    private final String LOG_tag = "YUYIN_FILEMANAGER";

    private ArrayList<FileItem> fileItemArrayList = new ArrayList<>();
    private FileAdapter adapter;
    private RecyclerView recyclerView;


    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        // Inflate the layout for this fragment
//        return inflater.inflate(R.layout.fragment_files_manager, container, false);
        binding = FragmentFilesManagerBinding.inflate(inflater, container, false);
        return binding.getRoot();
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        LinearLayoutManager linearLayoutManager = new LinearLayoutManager(getContext());
        recyclerView = binding.recyclerFile;
        recyclerView.setLayoutManager(linearLayoutManager);
        File dir_path = getContext().getExternalFilesDir(Environment.DIRECTORY_DOCUMENTS);
        File file = null;
        try {
            file = new File(dir_path.getCanonicalPath()+File.separator+"YuYin");
        } catch (IOException e) {
            e.printStackTrace();
        }
        File[] files = file.listFiles((dir, name) -> name.endsWith(".txt"));
        // 返回null
        if (files != null) {
            for (File item : files) {
                fileItemArrayList.add(new FileItem(item));
            }
        }
        adapter = new FileAdapter(fileItemArrayList);
        recyclerView.setAdapter(adapter);
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        binding = null;
    }
}

File: app/src/main/java/com/yuyin/demo/view/file/FilesManager.kt
package com.yuyin.demo.view.file

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import androidx.fragment.app.activityViewModels
import androidx.fragment.app.viewModels
import androidx.lifecycle.lifecycleScope
import androidx.navigation.Navigation
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.yuyin.demo.R
import com.yuyin.demo.databinding.FragmentFilesManagerBinding
import com.yuyin.demo.viewmodel.FilesManagerViewModel
import com.yuyin.demo.viewmodel.YuyinViewModel
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import java.io.File
import java.io.IOException
import com.yuyin.demo.utils.YuYinUtil.YuYinLog as Log


class FilesManager : Fragment() {
    val TAG = "FilesManager"
    private var binding: FragmentFilesManagerBinding? = null
    private val fileItemArrayList = ArrayList<FileItem>()
    private var adapter: FileAdapter? = null
    private var recyclerView: RecyclerView? = null
    val yuyinViewModel: YuyinViewModel by activityViewModels()
    val filesManagerViewModel: FilesManagerViewModel by viewModels()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        Log.i(TAG,"onCreate")
    }

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        Log.i(TAG,"onCreateView")
        binding = FragmentFilesManagerBinding.inflate(inflater, container, false)
        return binding!!.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        Log.i(TAG,"onViewCreated")
        super.onViewCreated(view, savedInstanceState)
        val linearLayoutManager = LinearLayoutManager(context)
        recyclerView = binding!!.recyclerFile
        recyclerView!!.layoutManager = linearLayoutManager
        val dir_path = yuyinViewModel.yuYinDataDir.toFile()
        val files = dir_path.listFiles { _: File?, name: String -> name.endsWith(".json") }
        // 返回null
        if (files != null) {
            for (item in files) {
                fileItemArrayList.add(FileItem(item))
            }
        }
        adapter = FileAdapter(fileItemArrayList,filesManagerViewModel,yuyinViewModel)
        recyclerView!!.adapter = adapter
        openFileListener()
    }

    override fun onDestroyView() {
        super.onDestroyView()
        Log.i(TAG,"onDestroyView")
        fileItemArrayList.clear()
        binding = null
    }

    override fun onDestroy() {
        super.onDestroy()
        Log.i(TAG,"onDestroy")
    }

    private fun openFileListener()  {
        viewLifecycleOwner.lifecycleScope.launch(Dispatchers.Main) {
            filesManagerViewModel.openFile.collect {
                Navigation.findNavController(
                    requireActivity(),
                    R.id.yuyin_nav_host_container_fragment
                ).navigate(R.id.action_filesManager_dest_to_edit_dest, Bundle().apply {
                    putString("file", it.absolutePath)
                })
            }
        }
    }
}